Live email preview when email settings are changed

NeosinneR opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Part of #52228. This PR adds option to see live changes in email preview when email settings are changed without needing to save settings. 

Closes #52424.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to **WooCommerce > Settings > Advanced > Features** and enable `Email improvements` experimental feature.
2. Go to **WooCommerce > Settings > Emails**.
3. Observe that the email preview is loaded.
4. Change (one by one) Logo, Header alignment, Footer text, and all five colors and observe that each change is reflected in the email preview. 
5. If your test website can send emails, check that the changes are also reflected in the email.
6. Reload the page and observe that the email preview is back to original values.
7. Test, that unless settings are saved, these styles are only applied in the email preview and not in real emails sent to customers.
    - This can be tested by changing the email styles, leaving settings tab open, and in a new tab resending an order email from order detail page. 

<!-- End testing instructions -->

NeosinneR committed:
Allow overriding email styles with transient values when in preview mode

NeosinneR committed:
Enable preview mode in EmailPreview class

NeosinneR committed:
Add new REST API route to save transients for email preview

NeosinneR committed:
Extract email preview iframe into separate component

NeosinneR committed:
Trigger "change" event on iris color picker fields to allow listening for change

NeosinneR committed:
Add change listeners for all email settings fields

NeosinneR committed:
Save transient value when email setting value changes

NeosinneR committed:
Add debounce, because color picker fires ton of events

NeosinneR committed:
Force iframe reload when email settings change

NeosinneR committed:
Sanitize footer text as textarea to preserve new lines

NeosinneR committed:
Live preview logo and footer text

NeosinneR committed:
Add changelog

NeosinneR committed:
Add REST API tests for save-transient route

NeosinneR committed:
Add test for using transients in email preview

NeosinneR committed:
Show experimental flag in list of Features

NeosinneR committed:
Add e2e test for live preview of email settings

NeosinneR committed:
Ensure that empty values (no logo or no footer) are reflected in live email preview

NeosinneR commented:
@alex-mpoet even though two jobs are failing, it's unrelated to this PR so it shouldn't block you from review.

- The `Build Live Branch` fails for everyone now. Reported in Slack: p1733957726482529-slack-C03CPM3UXDJ.
- The `Highlight Template Changes` is a bug, which I reported in #53646 and fixed in #53647. It's a false negative, so shouldn't prevent merging.

NeosinneR committed:
Delete email transients on page load to prevent conflicts with previous session

NeosinneR committed:
Fix lint error

alex-mpoet commented on plugins/woocommerce/tests/php/src/Internal/Admin/EmailPreview/EmailPreviewTest.php:
```diff
@@ -221,4 +221,18 @@ public function test_prepare_email_for_preview_filter() {
+	 * Test that transient values are applied in email preview
+	 */
+	public function test_transient_values_in_preview() {
+		update_option( EmailPreview::EMAIL_SETTINGS_IDS[0], 'option_value' );
```
Shouldn't we restore this option to its original value at the end? Or is it cleaned up somewhere globally after each test?

NeosinneR replied:
Good catch. Fixed in 8e823b4.

alex-mpoet committed:
Use jQuery change handlers for Undo button to unify it with preview handlers

alex-mpoet commented on plugins/woocommerce/client/admin/client/settings-email/settings-email-preview-iframe.tsx:
```diff
@@ -2,15 +2,48 @@
+			settingsIds.forEach( ( id ) => {
+				const field = jQuery( `#${ id }` );
+				if ( field.length ) {
+					field.off( 'change' );
```
This might detach other handlers not assigned by this component.

NeosinneR replied:
Yeah, as we discussed in Slack, for some reason, eslint is not happy with any of the approaches to remove a specific listener. I note this for a cleanup at the end of the project to see if we can get this to work properly, but hopefully it won't cause issues by then.

alex-mpoet reviewed (commented):
I have a few notes: 

1. The "Undo changes" button in Email colors isn't working. Presumably because there's a conflict between jQuery event handlers used by the preview and native handlers used by the Reset/Resync/Undo component. The latter could be modified to use jQuery handlers as well to fix this. Fixed in https://github.com/woocommerce/woocommerce/pull/53627/commits/879d510b6bf42a222373f4309f558c79f7f2591a.
2. After fixing the Undo button, there seems to be another issue. If you modify several colors and then click "Undo changes", only one color gets undone in the preview, the others stay changed. Same with "Sync with theme". 

I'll also leave these as non-actionable items:
1. Font family change handling is not implemented yet.
2. Email header/footer templates now use conditions to display different values depending of the preview flag, this means preview won't work in customized emails which use their own header/footer templates.

NeosinneR committed:
Fix undo/reset/sync button not working in email preview

NeosinneR commented:
@alex-mpoet thanks for fixing point 1. I fixed 2.  in c25e0ea.

NeosinneR committed:
Restore original value after test

alex-mpoet commented on plugins/woocommerce/client/admin/client/settings-email/settings-email-preview-iframe.tsx:
```diff
@@ -0,0 +1,69 @@
+					data: { key, value },
+				} );
+			} finally {
+				setIsLoading( false );
```
This makes the loading indicator disappear too early, before the frame starts loading. As a result, several seconds might pass between the end of the loading indication and the actual frame refreshing, which may leave an impression that changes aren't applied. 

I think this line can be removed, so it will be triggered by the iframe's `onLoad`. But if this is needed for error handling, replacing `finally` with `catch` or making this line conditional (only executed if an error occurs) could be a solution.

NeosinneR replied:
Great observation. This is indeed unnecessary at this place because iframe's onLoad will handle it. Removed in 4e1d407.

NeosinneR committed:
Remove unnecessary setIsLoading(false), which is handled by iframe onLoad

alex-mpoet commented on plugins/woocommerce/src/Internal/Admin/EmailPreview/EmailPreviewRestController.php:
```diff
@@ -98,6 +97,49 @@ public function register_routes() {
+							'sanitize_callback' => function ( $value, $request ) {
+								$key = $request->get_param( 'key' );
+								if ( 'woocommerce_email_footer_text' === $key ) {
+									return sanitize_textarea_field( $value );
```
If I have a `<br />` tag in my footer text, it works fine on the first page load, but disappears (no line break, two lines become one) if I do any changes to the footer text. It looks like the tag gets lost when the transient is saved.

NeosinneR replied:
Fixed in 808996f. There were different sanitization functions when saving option and transient.

NeosinneR committed:
Unify sanitization function when saving transient email settings

https://github.com/woocommerce/woocommerce/blob/ea734d756c518e3f59c5fa758bf86ef3e6b99eab/plugins/woocommerce/includes/admin/class-wc-admin-settings.php#L899

alex-mpoet reviewed (approved):
Thank you for the fixes! I have no more notes, things look fine!
